Add controllers

LeonovichMariia opened the pull request.

LeonovichMariia committed:
feat: добавлены модели User, UserDto, Item, ItemDto, Booking, BookingDto, ItemRequest, ItemRequestDto. Добавлены классы ValidationMessages и BookingStatus

LeonovichMariia committed:
feat: добавлены мапперы для User и для Item

LeonovichMariia committed:
feat: добавлены классы и интефейсы сервиса и хранилища для User и для Item а реализована логика в них. Добавлены классы ошибок, LogMessages и ErrorHandler

LeonovichMariia committed:
fix: code style

LeonovichMariia committed:
fix: code style

KirPyt commented on src/main/java/ru/practicum/shareit/booking/BookingStatus.java:
```diff
@@ -0,0 +1,8 @@
+package ru.practicum.shareit.booking;
+
+public enum BookingStatus {
```
Рекомендую добавить поле для хранения описания статуса бронирования

KirPyt commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,23 @@
+    private LocalDateTime start;
+    @FutureOrPresent(message = ValidationMessages.END_DATA)
+    private LocalDateTime end;
+    private String description;
```
От данного поля рекомендую отказаться, описание стоит хранить в перечислении в `BookingStatus`

KirPyt commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -0,0 +1,24 @@
+    private long id;
+    private LocalDateTime start;
+    private LocalDateTime end;
+    private String description;
```
Аналогично, рекомендую отказаться от данного поля и хранить описание состояния бронирования в `BookingStatus`

KirPyt commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,28 @@
+
+import ru.practicum.shareit.item.model.Item;
+
+public class ItemMapper {
```
Рекомендую добавить приватный конструктор, т.к данный класс служебный.
Можно сделать это через аннотацию

KirPyt commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,21 @@
+    private String name;
+    private String description;
+    private Boolean available;
+    private Long owner;
```
Здесь лучше хранить непосредственно `User`

KirPyt commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,26 @@
+    private String description;
+    @NotNull(message = ValidationMessages.AVAILABLE_NULL)
+    private Boolean available;
+    private Long owner;
```
Здесь лучше хранить непосредственно `User`

KirPyt commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+            log.warn(LogMessages.EMPTY_DESCRIPTION.toString(), itemDto);
+            throw new ValidationException(LogMessages.EMPTY_DESCRIPTION.toString());
+        }
+        if (itemDto.getName() == null || itemDto.getName().equals("") || itemDto.getName().isEmpty()) {
```
При наличии `isEmpty()` в `equals("")` нет смысла

KirPyt commented on src/main/java/ru/practicum/shareit/item/storage/ItemStorageImpl.java:
```diff
@@ -0,0 +1,101 @@
+    private Long id = 1L;
+    private final UserStorage userStorage;
+    private final Map<Long, Item> items = new HashMap<>();
+    private final Map<Long, Set<Long>> userPersonalItems = new HashMap<>();
```
Отличное решение хранить данную связку пользователей и вещей

KirPyt commented on src/main/java/ru/practicum/shareit/user/storage/UserStorageImpl.java:
```diff
@@ -0,0 +1,89 @@
+public class UserStorageImpl implements UserStorage {
+    private Long id = 1L;
+    private final Map<Long, User> users = new HashMap<>();
+
```
Предлагаю в репозиторий пользователей добавить `Set`, в котором будет храниться набор уникальных `email`, это обеспечит уникальность и позволит легко проверить наличие пользователя с таким `email`, это обеспечит проверку наличия `email` за сложность `О(1)`

KirPyt reviewed (commented):
Всё отлично)
Предлагаю поработать ещё над проектом, а именно - поправить мелкие замечания и попробовать учесть советы

LeonovichMariia committed:
delete: из классов Booking и BookingDto удалено поле с описанием.
feat: в BookingStatus добавлено поле для хранения описания статуса бронирования. В класс UserStorageImpl добавлен Set с электронными почтами пользователей и внесены соответствующие правки в логику.
fix: в ItemMapper и UserMapper добавлен приватный конструктор. У поля owner в классах Item и ItemDto изменен тип данных с Long на User

LeonovichMariia committed:
fix: code style
